### PR TITLE
Add Shelly gen2 authentication support

### PR DESCRIPTION
**Pull request by thecode**
Implement Shelly gen2 authentication support, continue work by @figadore https://github.com/home-assistant-libs/aioshelly/pull/192

Thanks @figadore and @kzyapkov

### CODE REVIEW

File: aioshelly/wsrpc.py [hunk 1]
```diff
@@ -25,26 +29,68 @@
...
+    """RPC Auth data class."""
+
+    realm: str
+    username: str
+    password: str
+
+    def __post_init__(self) -> None:
+        """Call after initialization."""
+        self.ha1 = hex_hash(f"{self.username}:{self.realm}:{self.password}")
+        self.ha2 = hex_hash("dummy_method:dummy_uri")
```
Comment (balloob) [inline aioshelly/wsrpc.py:48]: why generate this on `__post_init__` and not just statically set it? 

Let's add a comment why this is necessary.

File: aioshelly/wsrpc.py [hunk 2]
```diff
@@ -206,34 +258,62 @@ def connected(self) -> bool:
...
+
+        if code != 401:
+            raise JSONRPCError(code, msg)
+        if self._auth_data is None:
+            raise InvalidAuthError(code, msg)
+
+        # Update auth from response and try with new auth data
+        auth = json.loads(msg)
+        self._session.auth = self._auth_data.get_auth(auth["nonce"], auth.get("nc", 1))
+        resp = await self._rpc_call(method, params, timeout)
```
Comment (balloob) [inline aioshelly/wsrpc.py:281]: Better to call yourself again but now set a keyword arg to not refresh auth. Easier to keep the code in sync.

```suggestion
        return await self.call(method, params, timeout, handle_auth=False)
```

**Review by balloob — APPROVED** [review]

File: aioshelly/rpc_device.py [hunk 3]
```diff
@@ -88,13 +88,22 @@ async def initialize(self) -> None:
...
-                    self.update_device_info(),
-                    self.update_config(),
-                    self.update_status(),
+            if self.requires_auth:
+                if not self.options.auth:
+                    raise AuthRequired
+
+                self._wsrpc.set_auth_data(
+                    self.shelly["auth_domain"],
+                    cast(str, self.options.username),
```
Comment (balloob) [inline aioshelly/rpc_device.py:97]: Why does this need casting?

**Review by balloob — APPROVED** [review]

File: aioshelly/rpc_device.py (hunk 3 above)
Comment (thecode) [reply aioshelly/rpc_device.py]: I used the `self.options.auth`, this is why it confuses mypy, but there is no reason to use it here, changed it to 
https://github.com/home-assistant-libs/aioshelly/blob/00b286b4997ac176633d64b44ac2a51cf7273d3b/aioshelly/rpc_device.py#L92

**Review by balloob — APPROVED** [review]

**Pull request merged**